feat: Add coordinator implementation of CLP connector.

wraymo opened the pull request:
<!-- markdownlint-disable MD012 -->

<!--
Set the PR title to a meaningful commit message that:

* is in imperative form.
* follows the Conventional Commits specification (https://www.conventionalcommits.org).
  * See https://github.com/commitizen/conventional-commit-types/blob/master/index.json for possible
    types.

Example:

fix: Don't add implicit wildcards ('*') at the beginning and the end of a query (fixes #390).
-->

# Description

<!-- Describe what this request will change/fix and provide any details necessary for reviewers. -->
This PR introduces a CLP connector. The background and proposed implementation details are outlined in the associated [RFC](https://github.com/prestodb/rfcs/pull/37).

This PR implements one part of phase 1 of the proposed implementation, namely the Java implementation for the coordinator. The worker implementation will leverage **Velox** as the default query engine, so once the Velox PR is merged, we will submit another PR to this repo to add the necessary changes to `presto-native-execution`. 

Like other connectors, we have created a `presto-clp` module and implemented all required connector interfaces. The plan optimizer will be a future PR.

The important classes in the connector are described below.

---

### Core Classes in Java

#### `ClpConfig`
The configuration class for CLP. Currently, we support the following properties:

- **`clp.metadata-expire-interval`**: Defines the time interval after which metadata entries are considered expired and removed from the cache.
- **`clp.metadata-refresh-interval`**: Specifies how frequently metadata should be refreshed from the source to ensure up-to-date information.
- **`clp.polymorphic-type-enabled`**: Enables or disables support for polymorphic types within CLP. This determines whether dynamic type resolution is allowed.
- **`clp.metadata-provider-type`**: Defines the type of the metadata provider. It could be a database, a file-based store, or another external system. By default, we use MySQL.
- **`clp.metadata-db-url`**: The connection URL for the metadata database, used when `clp.metadata-provider-type` is configured to use a database.
- **`clp.metadata-db-name`**: The name of the metadata database.
- **`clp.metadata-db-user`**: The database user with access to the metadata database.
- **`clp.metadata-db-password`**: The password for the metadata database user.
- **`clp.metadata-table-prefix`**: A prefix applied to table names in the metadata database.
- **`clp.split-provider-type`**: Defines the type of split provider for query execution. By default, we use MySQL, and the connection parameters are the same as those for the metadata database.

---

#### `ClpSchemaTree`
A helper class for constructing a nested schema representation from CLP’s column definitions. It supports hierarchical column names (e.g., `a.b.c`), handles name/type conflicts when the `clp.polymorphic-type-enabled` option is enabled, and maps serialized CLP types to Presto types. The schema tree produces a flat list of `ClpColumnHandle` instances, including `RowType` for nested structures, making it suitable for dynamic or semi-structured data formats.

When polymorphic types are enabled, conflicting fields are given unique names by appending a type-specific suffix to the column name. For instance, if an integer field named "a" and a `Varstring` (CLP type) field named "a" coexist in CLP’s schema tree, they are represented as `a_bigint` and `a_varchar` in Presto. This approach ensures that such fields remain queryable while adhering to Presto’s constraints.

---

#### `ClpMetadataProvider`
An interface responsible for retrieving metadata from a specified source. 
```java
public interface ClpMetadataProvider {
    List<ClpColumnHandle> listColumnHandles(SchemaTableName schemaTableName);  
    List<ClpTableHandle> listTableNames(String schema);  
}

```
We provide a default implementation called `ClpMySqlMetadataProvider`, which uses two MySQL tables. One of these is the datasets table, defined with the schema shown below. Currently, we support only a single Presto schema named `default`, and this metadata table stores all table names, paths, and storage types associated with that Presto schema.

| Column Name  | Data Type       | Constraints   |
|--------------|-----------------|---------------|
| `name` | `VARCHAR(255)`  | `PRIMARY KEY` |
| `archive_storage_type` | `VARCHAR(4096)` | `NOT NULL`    |
| `archive_storage_directory` | `VARCHAR(4096)` | `NOT NULL`    |

The second MySQL table contains column metadata, defined by the schema shown below. Each Presto table is associated with a corresponding MySQL table that stores metadata about its columns.

| Column Name | Data Type   | Constraints             |
|------------|------------|------------------------|
| `name`     | `VARCHAR(512)` | `NOT NULL` |
| `type`     | `TINYINT`      | `NOT NULL` |
| **Primary Key** | (`name`, `type`) |  |

---

#### `ClpSplitProvider`
In CLP, an archive is the fundamental unit for searching, and we treat each archive as a Presto Split. This allows independent parallel searches across archives. The `ClpSplitProvider` interface, shown below, defines how to retrieve split information from a specified source:

```java
public interface ClpSplitProvider {
    List<ClpSplit> listSplits(ClpTableLayoutHandle clpTableLayoutHandle);  
}
```
We provide a default implementation called `ClpMySqlSplitProvider`. It uses an archive table to store archive IDs associated with each table. The table below shows part of the schema (some irrelevant fields are omitted).
| Column Name | Data Type   | Constraints             |
|------------|------------|------------------------|
| `pagination_id`     | `BIGINT` | `AUTO_INCREMENT PRIMARY KRY` |
|`id` | `VARCHAR(128)` | `NOT NULL` |
|...|...|...|

By concatenating the table path (`archive_storage_directory`) and the archive ID (`id`), we can retrieve all split paths for a table.

---

#### `ClpMetadata`
This interface enables Presto to access various metadata. All requests are delegated to `ClpMetadataProvider`

For metadata management, it also maintains two caches and periodically refreshes the metadata.
- **`columnHandleCache`**: A `LoadingCache<SchemaTableName, List<ClpColumnHandle>>` that maps a `SchemaTableName` to its corresponding list of `ClpColumnHandle` objects.
- **`tableHandleCache`**: A `LoadingCache<String, List<ClpTableHandle>>` that maps a schema name (`String`) to a list of `ClpTableHandle` objects


# Checklist

<!-- Ensure each item below is satisfied and indicate so by inserting an `x` within each `[ ]`. -->

* [x] The PR satisfies the [contribution guidelines][yscope-contrib-guidelines].
* [x] This is a breaking change and that has been indicated in the PR title, OR this isn't a
  breaking change.
* [x] Necessary docs have been updated, OR no docs need to be updated.

# Validation performed

<!-- Describe what tests and validation you performed on the change. -->
+ All unit tests passed successfully.
+ End-to-end tests verified, including Velox workers and Prestissimo protocol integration.

[yscope-contrib-guidelines]: https://docs.yscope.com/dev-guide/contrib-guides-overview.html


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced the CLP Connector for querying CLP-S archives in Presto with support for MySQL metadata and split providers.
  - Added support for nested and polymorphic column types, mapping CLP data types to Presto types including complex ROW structures.
  - Provided extensive configuration options for metadata refresh, expiration intervals, polymorphic type support, and provider selection.
  - Integrated lifecycle management and seamless plugin framework support within Presto.

- **Tests**
  - Added tests covering metadata and split provider functionality, validating schema, table, and split retrieval.
  - Included utilities for setting up and tearing down H2-based test metadata databases.

- **Documentation**
  - Delivered comprehensive CLP Connector documentation detailing setup, configuration, data type mappings, and usage.
  - Updated connector index to include the new CLP Connector.

- **Chores**
  - Updated build and provisioning configurations to incorporate the new CLP Connector module and plugin packaging.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

wraymo commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/metadata/ClpSchemaTree.java:
```diff
@@ -0,0 +1,199 @@
+    public List<ClpColumnHandle> collectColumnHandles()
+    {
+        List<ClpColumnHandle> columns = new ArrayList<>();
+        for (Map.Entry<String, ClpNode> entry : root.children.entrySet()) {
+            String name = entry.getKey();
+            ClpNode child = entry.getValue();
+            if (child.isLeaf()) {
+                columns.add(new ClpColumnHandle(name, child.originalName, child.type, true));
+            }
+            else {
+                Type rowType = buildRowType(child);
+                columns.add(new ClpColumnHandle(name, child.originalName, rowType, true));
+            }
```
The order doesn't matter

wraymo commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/metadata/ClpMetadataProvider.java:
```diff
@@ -0,0 +1,33 @@
+
+import java.util.List;
+
+public interface ClpMetadataProvider
+{
+    /**
+     * Returns the list of column handles for the given table.
+     */
+    List<ClpColumnHandle> listColumnHandles(SchemaTableName schemaTableName);
+
+    /**
+     * Returns the list of table handles in the given schema.
+     */
+    List<ClpTableHandle> listTableHandles(String schema);
+}
```
No, it's used by the coordinator, and there's only one instance

wraymo commented on presto-clp/src/test/java/com/facebook/presto/plugin/clp/TestClpMetadata.java:
```diff
@@ -0,0 +1,131 @@
+        ConnectorTableMetadata tableMetadata = metadata.getTableMetadata(SESSION, clpTableHandle);
+        HashSet<ColumnMetadata> columnMetadata = new HashSet<>();
+        columnMetadata.add(ColumnMetadata.builder()
+                .setName("a_bigint")
+                .setType(BigintType.BIGINT)
+                .setNullable(true)
+                .build());
+        columnMetadata.add(ColumnMetadata.builder()
+                .setName("a_varchar")
+                .setType(VarcharType.VARCHAR)
+                .setNullable(true)
+                .build());
+        columnMetadata.add(ColumnMetadata.builder()
+                .setName("b_double")
+                .setType(DoubleType.DOUBLE)
+                .setNullable(true)
+                .build());
+        columnMetadata.add(ColumnMetadata.builder()
+                .setName("b_varchar")
+                .setType(VarcharType.VARCHAR)
+                .setNullable(true)
+                .build());
+        columnMetadata.add(ColumnMetadata.builder()
+                .setName("c")
+                .setType(RowType.from(ImmutableList.of(
+                        RowType.field("d", BooleanType.BOOLEAN),
+                        RowType.field("e", VarcharType.VARCHAR))))
+                .setNullable(true)
+                .build());
+        columnMetadata.add(ColumnMetadata.builder()
+                .setName("f")
+                .setType(RowType.from(ImmutableList.of(
+                        RowType.field("g",
+                                RowType.from(ImmutableList.of(
+                                        RowType.field("h", new ArrayType(VarcharType.VARCHAR))))))))
+                .setNullable(true)
+                .build());
+        assertEquals(columnMetadata, new HashSet<>(tableMetadata.getColumns()));
+    }
+}
```
`org.apache.commons.lang3.tuple.Pair` is abstract.

anlowee commented on presto-clp/pom.xml:
```diff
@@ -49,12 +49,6 @@
-            <artifactId>jsr305</artifactId>
-            <optional>true</optional>
-        </dependency>
-
```
why?

anlowee commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/ClpMetadata.java:
```diff
@@ -0,0 +1,199 @@
+            return ImmutableList.of();
+        }
+
+        return listTables(schemaNameValue).stream()
```
so it seems the `tableHandle` is just to get the table name. Why not directly cache the table name?

anlowee commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/ClpExpression.java:
```diff
@@ -0,0 +1,70 @@
+    // The remaining (non-translatable) portion of the RowExpression, if any.
+    private final Optional<RowExpression> remainingExpression;
+
+    public ClpExpression(Optional<String> definition, Optional<RowExpression> remainingExpression)
```
There are some IDEA warnings about that the `Optional` should only be used in the resturn type not the argument type. I think it make sense. Maybe in the caller we can call something like `new ClpExpression(definition.orElse(null), remainingExpression.orElse(null))` instead of definiing it as `Opational` here.

wraymo replied:
But I think it's pretty common in Presto code. e.g. `BasicQueryInfo`, `QueryStateInfo`, `SessionRepresentation`, etc.

kirkrodrigues replied:
Read a bit [about it](https://github.com/SwingGuy1024/Blog/blob/master/Bad%20Uses%20of%20Optional.md). I don't think there's a good reason to use `Optional` for parameters.

However, Presto's [style guide](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#code-style) does require it for public method parameters if the method isn't performance critical.

Personally, I would just use overloading in this case (i.e., remove this constructor and get rid of the Optional in the other constructors).

anlowee replied:
Although this class is delayed to the next PR, I think this constructor can be considered as performance critical because it is in the critical path (to generate the pushdown and metadata filter SQL). So I think we can use `@Nullable` to annotate the nullable parameters.

wraymo replied:
This is not in the critical path, as there are over 100 plan optimizers, and in most real-world cases, the coordinator accounts for less than 10%—or even under 1%—of the total query time. Even if all of the coordinator's time were spent on plan optimization, its impact on the overall query latency would be below 0.1%. The true critical path lies in data scanning and expression evaluation. As discussed before, our plan was to pass raw types but store them as `Optional`

anlowee commented on presto-clp/src/test/java/com/facebook/presto/plugin/clp/ClpMetadataDbSetUp.java:
```diff
@@ -0,0 +1,169 @@
+public final class ClpMetadataDbSetUp
+{
+    public static final String metadataDbUrlTemplate =
+            "jdbc:h2:file:/tmp/%s;MODE=MySQL;DATABASE_TO_UPPER=FALSE";
```
and add a method to get this path, so that it can be used in the subsequent unit tests.

anlowee replied:
As discussed offline, will fix in the future.

anlowee replied:
@coderabitai Can you help to create an issue to track this?

anlowee commented on presto-clp/src/test/java/com/facebook/presto/plugin/clp/ClpMetadataDbSetUp.java:
```diff
@@ -0,0 +1,169 @@
+        File lockFile = new File(String.format("/tmp/%s.trace.db", dbName)); // Optional, H2 sometimes creates this
+        if (dbFile.exists()) {
+            dbFile.delete();
+            System.out.println("Deleted database file: " + dbFile.getAbsolutePath());
```
Remove this or log it.

anlowee commented on presto-clp/src/test/java/com/facebook/presto/plugin/clp/ClpMetadataDbSetUp.java:
```diff
@@ -0,0 +1,169 @@
+
+import static org.testng.Assert.fail;
+
+public final class ClpMetadataDbSetUp
```
Do we want to rename it start with `Test`?

wraymo replied:
No, it's not a unit test

anlowee commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/ClpMetadata.java:
```diff
@@ -0,0 +1,199 @@
+    }
+
+    @Override
+    public List<ConnectorTableLayoutResult> getTableLayouts(ConnectorSession session,
```
It says this function is deprecated:
```
/**
     * Return a list of table layouts that satisfy the given constraint.
     * <p>
     * For each layout, connectors must return an "unenforced constraint" representing the part of the constraint summary that isn't guaranteed by the layout.
     *
     * @deprecated replaced by {@link ConnectorMetadata#getTableLayoutForConstraint(ConnectorSession, ConnectorTableHandle, Constraint, Optional)}
     */
```

anlowee replied:
are we going to address this or ignore this for any reasons?

kirkrodrigues commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/ClpConfig.java:
```diff
@@ -0,0 +1,170 @@
+    private long metadataExpireInterval = 600;
+    private SplitProviderType splitProviderType = SplitProviderType.MYSQL;
+
+    public static final Pattern SAFE_SQL_IDENTIFIER = Pattern.compile("^[a-zA-Z0-9_]+$");
```
How about `SAFE_SQL_TABLE_NAME_PATTERN`?

kirkrodrigues commented on presto-docs/src/main/sphinx/connector/clp.rst:
```diff
@@ -0,0 +1,178 @@
+Overview
+--------
+
+The CLP Connector enables SQL-based querying of CLP-S archives from Presto. This document describes how to set up the
+CLP Connector to run SQL queries.
```
```suggestion
The CLP Connector enables SQL-based querying of `CLP <https://github.com/y-scope/clp>`_ archives via Presto. This
document describes how to configure the CLP Connector for use with a CLP cluster, as well as essential details for
understanding the CLP connector.
```

kirkrodrigues commented on presto-docs/src/main/sphinx/connector/clp.rst:
```diff
@@ -0,0 +1,178 @@
+Configuration
+-------------
+
+To configure the CLP connector, create a catalog properties file
+``etc/catalog/clp.properties`` with the following contents,
+replacing the properties as appropriate:
```
```suggestion
To configure the CLP connector, create a catalog properties file ``etc/catalog/clp.properties`` with at least the
following contents, modifying the properties as appropriate:
```

kirkrodrigues commented on presto-docs/src/main/sphinx/connector/clp.rst:
```diff
@@ -0,0 +1,178 @@
+``etc/catalog/clp.properties`` with the following contents,
+replacing the properties as appropriate:
+
+.. code-block:: none
```
```suggestion
.. code-block:: ini
```

kirkrodrigues commented on presto-docs/src/main/sphinx/connector/clp.rst:
```diff
@@ -0,0 +1,178 @@
+================================== ======================================================================== =========
+Property Name                      Description                                                              Default
+================================== ======================================================================== =========
+``clp.metadata-expire-interval``   Defines how long metadata entries remain valid before being considered   600
+                                   expired, in seconds.
```
```suggestion
``clp.metadata-expire-interval``   Defines how long, in seconds, metadata entries remain valid before they  600
                                   need to be refreshed.
```

kirkrodrigues commented on presto-docs/src/main/sphinx/connector/clp.rst:
```diff
@@ -0,0 +1,178 @@
+                                   semi-structured data where the same field may appear with different
+                                   types.
+
+                                   When enabled, type annotations are added to conflicting field names to
+                                   distinguish between types. For example, if ``id`` column appears as both
+                                   an ``int`` and ``string`` types, the connector will create two columns
+                                   named ``id_bigint`` and ``id_varchar``.
```
```suggestion
                                   When enabled, type annotations are added to conflicting field names to
                                   distinguish between types. For example, if an ``id`` column appears with
                                   both an ``int`` and ``string`` types, the connector will create two
                                   columns named ``id_bigint`` and ``id_varchar``.
```

kirkrodrigues commented on presto-docs/src/main/sphinx/connector/clp.rst:
```diff
@@ -0,0 +1,178 @@
+================================== ======================================================================== =========
+
+
+
```
```suggestion
```

kirkrodrigues commented on presto-docs/src/main/sphinx/connector/clp.rst:
```diff
@@ -0,0 +1,178 @@
+
+Metadata and Split Providers
+----------------------------
+The CLP connector relies on metadata and split providers to retrieve information from various sources. By default, it
```
```suggestion

The CLP connector relies on metadata and split providers to retrieve information from various sources. By default, it
```

kirkrodrigues commented on presto-docs/src/main/sphinx/connector/clp.rst:
```diff
@@ -0,0 +1,178 @@
+uses a MySQL database for both metadata and split storage. We recommend using the CLP package for log ingestion, which
+automatically populates the database with the required information.
+
+If you prefer to use a different source—or the same source with a custom implementation—you can provide your own
```
```suggestion
If you prefer to use a different source--or the same source with a custom implementation--you can provide your own
```

kirkrodrigues commented on presto-docs/src/main/sphinx/connector/clp.rst:
```diff
@@ -0,0 +1,178 @@
+
+Object Types
+^^^^^^^^^^^^
+CLP stores metadata using a global schema tree structure that captures all possible fields from various log structures.
```
```suggestion

CLP stores metadata using a global schema tree structure that captures all possible fields from various log structures.
```

kirkrodrigues reviewed (changes requested):
Besides the inline comments, let's update the code to respect the Presto [contribution guidelines](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#code-style). Specifically:

* There are some files using the wrong docstring format like `ClpMetadataProvider`.
* The ordering of fields and methods, including their ordering by visibility disagrees with the guidelines.
* Not all files are fully formatted using the code style file they provide.

wraymo commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/metadata/ClpMySqlMetadataProvider.java:
```diff
@@ -0,0 +1,135 @@
+    private final ClpConfig config;
+
+    // Column names
+    private static final String COLUMN_METADATA_TABLE_COLUMN_NAME = "name";
+    private static final String COLUMN_METADATA_TABLE_COLUMN_TYPE = "type";
+    private static final String DATASETS_TABLE_COLUMN_NAME = "name";
+    private static final String DATASETS_TABLE_COLUMN_ARCHIVE_STORAGE_TYPE = "archive_storage_type";
+    private static final String DATASETS_TABLE_COLUMN_ARCHIVE_STORAGE_DIRECTORY = "archive_storage_directory";
+
+    // Table suffixes
+    private static final String COLUMN_METADATA_TABLE_SUFFIX = "_column_metadata";
+    private static final String DATASETS_TABLE_SUFFIX = "datasets";
+
+    // SQL templates
+    private static final String SQL_SELECT_COLUMN_METADATA_TEMPLATE = "SELECT * FROM `%s%s" +
+            COLUMN_METADATA_TABLE_SUFFIX + "`";
+    private static final String SQL_SELECT_DATASETS_TEMPLATE =
+            String.format("SELECT `%s`, `%s`, `%s` FROM `%%s%s`", DATASETS_TABLE_COLUMN_NAME,
+                    DATASETS_TABLE_COLUMN_ARCHIVE_STORAGE_TYPE, DATASETS_TABLE_COLUMN_ARCHIVE_STORAGE_DIRECTORY,
+                    DATASETS_TABLE_SUFFIX);
```
Sure

wraymo commented on presto-clp/src/test/java/com/facebook/presto/plugin/clp/ClpMetadataDbSetUp.java:
```diff
@@ -0,0 +1,229 @@
+    private static final String DATASETS_TABLE_NAME = METADATA_DB_TABLE_PREFIX + DATASETS_TABLE_SUFFIX;
+    private static final String ARCHIVE_TABLE_COLUMN_PAGINATION_ID = "pagination_id";
+
+    private ClpMetadataDbSetUp()
```
I think it should, but when I apply their code style xml, it autoformats like this

kirkrodrigues replied:
Gotcha.

kirkrodrigues commented on presto-docs/src/main/sphinx/connector/clp.rst:
```diff
@@ -0,0 +1,179 @@
+uses a MySQL database for both metadata and split storage. We recommend using the CLP package for log ingestion, which
+automatically populates the database with the required information.
+
+If you prefer to use a different source--or the same source with a custom implementation—you can provide your own
```
```suggestion
If you prefer to use a different source--or the same source with a custom implementation--you can provide your own
```

anlowee commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/ClpConnector.java:
```diff
@@ -0,0 +1,82 @@
+    @Override
+    public ConnectorTransactionHandle beginTransaction(IsolationLevel isolationLevel, boolean readOnly)
+    {
+        return ClpTransactionHandle.INSTANCE;
```
import the static `INSTANCE`

anlowee commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/ClpHandleResolver.java:
```diff
@@ -0,0 +1,55 @@
+public class ClpHandleResolver
+        implements ConnectorHandleResolver
+{
+    @Override
+    public Class<? extends ConnectorTableHandle> getTableHandleClass()
+    {
+        return ClpTableHandle.class;
+    }
+
+    @Override
+    public Class<? extends ConnectorTableLayoutHandle> getTableLayoutHandleClass()
+    {
+        return ClpTableLayoutHandle.class;
+    }
+
+    @Override
+    public Class<? extends ColumnHandle> getColumnHandleClass()
+    {
+        return ClpColumnHandle.class;
+    }
+
+    @Override
+    public Class<? extends ConnectorSplit> getSplitClass()
+    {
+        return ClpSplit.class;
+    }
+
+    @Override
+    public Class<? extends ConnectorTransactionHandle> getTransactionHandleClass()
+    {
+        return ClpTransactionHandle.class;
+    }
```
```suggestion
    @Override
    public Class<? extends ColumnHandle> getColumnHandleClass()
    {
        return ClpColumnHandle.class;
    }

    @Override
    public Class<? extends ConnectorTableHandle> getTableHandleClass()
    {
        return ClpTableHandle.class;
    }

    @Override
    public Class<? extends ConnectorSplit> getSplitClass()
    {
        return ClpSplit.class;
    }

    @Override
    public Class<? extends ConnectorTableLayoutHandle> getTableLayoutHandleClass()
    {
        return ClpTableLayoutHandle.class;
    }

    @Override
    public Class<? extends ConnectorTransactionHandle> getTransactionHandleClass()
    {
        return ClpTransactionHandle.class;
    }
```
Alphabize

anlowee commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/ClpMetadata.java:
```diff
@@ -0,0 +1,207 @@
+        SchemaTableName schemaTableName = clpTableHandle.getSchemaTableName();
+        List<ColumnMetadata> columns = listColumns(schemaTableName).stream()
+                .map(ClpColumnHandle::getColumnMetadata)
+                .collect(ImmutableList.toImmutableList());
```
directly use `toImmutableList` after staticly import it.

anlowee commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/ClpMetadata.java:
```diff
@@ -0,0 +1,207 @@
+
+        List<SchemaTableName> schemaTableNames;
+        if (prefix.getTableName() == null) {
+            schemaTableNames = listTables(session, Optional.ofNullable(prefix.getSchemaName()));
```
static import `Optional.ofNullable`

anlowee commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/ClpMetadata.java:
```diff
@@ -0,0 +1,207 @@
+        }
+        else {
+            SchemaTableName table = new SchemaTableName(schemaName, prefix.getTableName());
+            if (listTables(session, Optional.ofNullable(schemaName)).contains(table)) {
```
same

anlowee commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/ClpMetadata.java:
```diff
@@ -0,0 +1,207 @@
+    {
+        ClpTableHandle clpTableHandle = (ClpTableHandle) tableHandle;
+        return listColumns(clpTableHandle.getSchemaTableName()).stream()
+                .collect(ImmutableMap.toImmutableMap(
```
same

anlowee commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/ClpMetadata.java:
```diff
@@ -0,0 +1,207 @@
+        return clpMetadataProvider.listTableHandles(schemaName);
+    }
+
+    private List<ClpTableHandle> listTables(String schemaName)
+    {
+        return tableHandleCache.getUnchecked(schemaName);
+    }
+
+    private List<ClpColumnHandle> listColumns(SchemaTableName schemaTableName)
+    {
+        return columnHandleCache.getUnchecked(schemaTableName);
+    }
```
```suggestion
    private List<ClpColumnHandle> listColumns(SchemaTableName schemaTableName)
    {
        return columnHandleCache.getUnchecked(schemaTableName);
    }

    private List<ClpTableHandle> listTables(String schemaName)
    {
        return tableHandleCache.getUnchecked(schemaName);
    }
```
alphabize

anlowee commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/ClpModule.java:
```diff
@@ -0,0 +1,54 @@
+    @Override
+    protected void setup(Binder binder)
+    {
+        binder.bind(ClpConnector.class).in(Scopes.SINGLETON);
```
static import `SINGLETON`

anlowee replied:
then directly use it for this file

anlowee commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/ClpModule.java:
```diff
@@ -0,0 +1,54 @@
+        configBinder(binder).bindConfig(ClpConfig.class);
+
+        ClpConfig config = buildConfigObject(ClpConfig.class);
+        if (config.getMetadataProviderType() == ClpConfig.MetadataProviderType.MYSQL) {
```
Not sure for this case do we want to:
```
import static com.yourpkg.ClpConfig.MetadataProviderType;
import static com.yourpkg.ClpConfig.SplitProviderType;
```
Then use `MetadataProviderType.MYSQL` and `SplitProviderType.MYSQL` to follow the code-style's static import guide.

WDYT? @kirkrodrigues

anlowee commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/ClpModule.java:
```diff
@@ -0,0 +1,54 @@
+            binder.bind(ClpMetadataProvider.class).to(ClpMySqlMetadataProvider.class).in(Scopes.SINGLETON);
+        }
+        else {
+            throw new PrestoException(ClpErrorCode.CLP_UNSUPPORTED_METADATA_SOURCE, "Unsupported metadata provider type: " + config.getMetadataProviderType());
```
static import `CLP_UNSUPPORTED_METADATA_SOURCE`

anlowee commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/ClpModule.java:
```diff
@@ -0,0 +1,54 @@
+            binder.bind(ClpSplitProvider.class).to(ClpMySqlSplitProvider.class).in(Scopes.SINGLETON);
+        }
+        else {
+            throw new PrestoException(ClpErrorCode.CLP_UNSUPPORTED_SPLIT_SOURCE, "Unsupported split provider type: " + config.getSplitProviderType());
```
same

anlowee commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/ClpSplit.java:
```diff
@@ -0,0 +1,63 @@
+    @Override
+    public List<HostAddress> getPreferredNodes(NodeProvider nodeProvider)
+    {
+        return ImmutableList.of();
```
static import `of`

anlowee commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/ClpSplitManager.java:
```diff
@@ -0,0 +1,48 @@
+                                          ConnectorTableLayoutHandle layout,
+                                          SplitSchedulingContext splitSchedulingContext)
+    {
+        ClpTableLayoutHandle layoutHandle = (ClpTableLayoutHandle) layout;
+        return new FixedSplitSource(clpSplitProvider.listSplits(layoutHandle));
+    }
```
Are we going to fix in the subsequent PRs? If not shall we open an issue to track this? It is performance related.

anlowee commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/ClpTableHandle.java:
```diff
@@ -0,0 +1,97 @@
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(schemaTableName, tablePath, storageType);
```
static import `hash`

anlowee commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/ClpColumnHandle.java:
```diff
@@ -0,0 +1,117 @@
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(columnName, originalColumnName, columnType, nullable);
```
static import `hash`

anlowee commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/ClpColumnHandle.java:
```diff
@@ -0,0 +1,117 @@
+            return false;
+        }
+        ClpColumnHandle other = (ClpColumnHandle) obj;
+        return Objects.equals(this.columnName, other.columnName) &&
```
static import `equals`

anlowee commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/ClpTableHandle.java:
```diff
@@ -0,0 +1,97 @@
+            return false;
+        }
+        ClpTableHandle other = (ClpTableHandle) obj;
+        return this.schemaTableName.equals(other.schemaTableName) &&
+                this.tablePath.equals(other.tablePath) &&
+                this.storageType == other.storageType;
```
For other `equals` like `ClpColumnHandle` and `ClpTableLayoutHandle`, we use `Object.equals` to compare two properties but here we use `.equals`. It would be better to keep consistent.

If switch to `Object.equals` then we should static import it.

anlowee commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/ClpTableLayoutHandle.java:
```diff
@@ -0,0 +1,78 @@
+            return false;
+        }
+        ClpTableLayoutHandle that = (ClpTableLayoutHandle) o;
+        return Objects.equals(table, that.table) &&
```
static import `equals`

anlowee commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/ClpTableLayoutHandle.java:
```diff
@@ -0,0 +1,78 @@
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(table, kqlQuery);
```
static import `hash`

anlowee commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/metadata/ClpMySqlMetadataProvider.java:
```diff
@@ -0,0 +1,140 @@
+        catch (SQLException e) {
+            log.warn("Failed to load table names: %s", e);
+        }
+        return tableHandles;
```
This is to respect the code-style:
> Prefer Immutable collections in Guava when possible. For example, instead of using

anlowee commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/metadata/ClpSchemaTreeNodeType.java:
```diff
@@ -0,0 +1,68 @@
+    private static final ClpSchemaTreeNodeType[] LOOKUP_TABLE;
+    private final byte type;
+
+    ClpSchemaTreeNodeType(byte type)
```
public before constructor

anlowee commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/split/ClpMySqlSplitProvider.java:
```diff
@@ -0,0 +1,99 @@
+            log.warn("Database error while processing splits for %s: %s", tableName, e);
+        }
+
+        return splits;
```
consider use `copyof`

anlowee commented on presto-clp/src/test/java/com/facebook/presto/plugin/clp/ClpMetadataDbSetUp.java:
```diff
@@ -0,0 +1,224 @@
+
+    public static DbHandle getDbHandle(String dbName)
+    {
+        return new DbHandle(format("/tmp/presto-clp-test-%s/%s", UUID.randomUUID(), dbName));
```
static import `UUID.randomUUID`

anlowee commented on presto-clp/src/test/java/com/facebook/presto/plugin/clp/ClpMetadataDbSetUp.java:
```diff
@@ -0,0 +1,224 @@
+    {
+        public String dbPath;
+
+        DbHandle(String dbPath)
```
public before the constructor

anlowee commented on presto-clp/src/test/java/com/facebook/presto/plugin/clp/TestClpMetadata.java:
```diff
@@ -0,0 +1,130 @@
+                ImmutableMap.of(
+                        TABLE_NAME,
+                        ImmutableList.of(
+                                new Pair<>("a", ClpSchemaTreeNodeType.Integer),
+                                new Pair<>("a", ClpSchemaTreeNodeType.VarString),
+                                new Pair<>("b", ClpSchemaTreeNodeType.Float),
+                                new Pair<>("b", ClpSchemaTreeNodeType.ClpString),
+                                new Pair<>("c.d", ClpSchemaTreeNodeType.Boolean),
+                                new Pair<>("c.e", ClpSchemaTreeNodeType.VarString),
+                                new Pair<>("f.g.h", ClpSchemaTreeNodeType.UnstructuredArray))));
```
static import these types

anlowee commented on presto-clp/src/test/java/com/facebook/presto/plugin/clp/TestClpMetadata.java:
```diff
@@ -0,0 +1,130 @@
+    {
+        HashSet<SchemaTableName> tables = new HashSet<>();
+        tables.add(new SchemaTableName(DEFAULT_SCHEMA_NAME, TABLE_NAME));
+        assertEquals(new HashSet<>(metadata.listTables(SESSION, Optional.empty())), tables);
```
static import `Optional.empty`

anlowee commented on presto-clp/src/test/java/com/facebook/presto/plugin/clp/TestClpMetadata.java:
```diff
@@ -0,0 +1,130 @@
+                        .build())
+                .add(ColumnMetadata.builder()
+                        .setName("f")
+                        .setType(RowType.from(ImmutableList.of(
+                                RowType.field("g",
```
static import `RowType.from` and `RowType.field`

anlowee replied:
maybe ignore this one as well. Maybe in the future we can ignore the static variable/methods which are too general. For example these `of`, `from` etc. But for `toImmutableList` which has the `List` in the method name so it is not general, then we need to static import.

anlowee commented on presto-clp/src/test/java/com/facebook/presto/plugin/clp/TestClpMetadata.java:
```diff
@@ -0,0 +1,130 @@
+    @AfterMethod
+    public void tearDown()
+    {
+        ClpMetadataDbSetUp.tearDown(dbHandle);
```
static import `tearDown`

wraymo replied:
conflict with `tearDown`

anlowee commented on presto-docs/src/main/sphinx/connector/clp.rst:
```diff
@@ -0,0 +1,179 @@
+
+.. code-block:: ini
+
+    connector.name=clp
+    clp.metadata-provider-type=mysql
+    clp.metadata-db-url=jdbc:mysql://localhost:3306
+    clp.metadata-db-name=clp_db
+    clp.metadata-db-user=clp_user
+    clp.metadata-db-password=clp_password
+    clp.metadata-table-prefix=clp_
+    clp.split-provider-type=mysql
```
if we applied the alphabize comment in the `ClpConfig`, maybe we should also alphabize here

anlowee commented on presto-docs/src/main/sphinx/connector/clp.rst:
```diff
@@ -0,0 +1,179 @@
+================================== ======================================================================== =========
+Property Name                      Description                                                              Default
+================================== ======================================================================== =========
+``clp.metadata-expire-interval``   Defines how long, in seconds, metadata entries remain valid before they  600
```
same, we should keep the same order

wraymo commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/ClpPlugin.java:
```diff
@@ -0,0 +1,28 @@
+    @Override
+    public Iterable<ConnectorFactory> getConnectorFactories()
+    {
+        return ImmutableList.of(new ClpConnectorFactory());
```
That's pretty weird in this case

anlowee replied:
agree

anlowee replied:
maybe ignore this `of` comment

wraymo commented on presto-clp/src/main/java/com/facebook/presto/plugin/clp/ClpErrorCode.java:
```diff
@@ -0,0 +1,43 @@
+
+    private final ErrorCode errorCode;
+
+    ClpErrorCode(int code, ErrorType type)
```
You cannot use `public`, either no modifier or explicitly `private`.

anlowee replied:
Is it only for enum class or for all? If it is for all maybe we should remove the `public` for those constructors already have one.

wraymo commented:
@anlowee I’ve updated the PR. There are a few cases I intentionally left unchanged:

1. For general method names like `of`, `copyOf`, `equals`, or symbols such as `RowType.fromType` that are not typically used as static imports in PrestoDB, I avoided static imports. Using the qualified form improves readability by preventing ambiguity with overly generic names.
2. For `tearDown`, I added a qualifier to clearly distinguish it from the class's own `tearDown` method.
3. The private modifier on enum constructors is redundant, so I removed it.
4. For enum comparisons, using `==` is preferred over `.equals()` for clarity and correctness.
